pkg/compactor/compactor.go

```diff
@@ -458,6 +464,35 @@ func (c *MultitenantCompactor) starting(ctx context.Context) error {
 	return nil
 }
 
+func newRingAndLifecycler(cfg RingConfig, logger log.Logger, reg prometheus.Registerer) (*ring.Ring, *ring.BasicLifecycler, error) {
+	kvStore, err := kv.NewClient(cfg.KVStore, ring.GetCodec(), kv.RegistererWithKVName(reg, "compactor-lifecycler"), logger)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to initialize compactors' KV store")
+	}
+
+	lifecyclerCfg, err := cfg.ToBasicLifecyclerConfig(logger)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to build compactors' lifecycler config")
+	}
+
+	var delegate ring.BasicLifecyclerDelegate
+	delegate = ring.NewInstanceRegisterDelegate(ring.ACTIVE, lifecyclerCfg.NumTokens)
+	delegate = ring.NewLeaveOnStoppingDelegate(delegate, logger)
+	delegate = ring.NewAutoForgetDelegate(ringAutoForgetUnhealthyPeriods*lifecyclerCfg.HeartbeatTimeout, delegate, logger)
+
+	compactorsLifecycler, err := ring.NewBasicLifecycler(lifecyclerCfg, "compactor", ringKey, kvStore, delegate, logger, prometheus.WrapRegistererWithPrefix("cortex_", reg))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to initialize compactors' lifecycler")
+	}
+
+	compactorsRing, err := ring.NewWithStoreClientAndStrategy(cfg.ToRingConfig(), "compactor", ringKey, kvStore, ring.NewDefaultReplicationStrategy(), prometheus.WrapRegistererWithPrefix("cortex_", reg), logger)
```

Comment:
We're reusing `kvStore` for the ring client too. `kvStore` is created with name `compactor-lifecycler` which doesn't apply if we reuse it for the ring client too. In the old implementation, the name of the KV store used for the client was called `compactor-ring`. I see two options:
1. Use two different KV stores, to keep previous behaviour
2. Rename `kvStore` from `compactor-lifecycler` to just `compactor`, then please double check if there's any dashboard or alert to update

Reply:
I will take a look at this.

Reply:
I'm going to go back to use two different KV stores and maintain the previous behavior rather than have compactors using a different setup than other components.

Reply:
This also exposes a problem with the way the kvstore is created for compactors (via this PR) and distributors: The `prometheus.Registerer` used for the kvstore isn't prefixed with `cortex_` so it emits metrics like `kv_request_duration_seconds_count` compared to ingesters which use `cortex_kv_request_duration_seconds_count`